File: scala-flink/src/main/scala/com/example/WordCount.scala
package com.example
import org.apache.flink.api.scala.ExecutionEnvironment
import org.apache.flink.api.scala._

object WordCount {
  def main(args: Array[String]): Unit = {
    // Set up the execution environment
    val env: ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment

    // Input data (static)
    val text = env.fromElements(
      "Apache Flink is a framework and distributed processing engine",
      "Flink provides batch processing and data stream processing",
      "Word count example using Flink"
    )

    // Transform the input data
    val wordCounts = text
      .flatMap(_.toLowerCase.split("\\W+"))  // Split lines into words
      .map((_, 1))                           // Convert words to (word, 1)
      .groupBy(0)                            // Group by the word
      .sum(1)                                // Sum the occurrences

    // Print the result
    wordCounts.print()
  }
}

File: flink-sample/WordCount.java


import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

public class WordCount {
    public static void main(String[] args) throws Exception {
        // Set up the streaming execution environment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // Define the source: read from a socket
        String hostname = "localhost";
        int port = 9999;

        // Get input data by connecting to the socket
        env.socketTextStream(hostname, port)
           .flatMap(new Tokenizer())
           .keyBy(value -> value.f0)
           .sum(1)
           .print();

        // Execute the program
        env.execute("Socket Window WordCount");
    }

    // User-defined function to split lines into words
    public static final class Tokenizer implements FlatMapFunction<String, Tuple2<String, Integer>> {
        @Override
        public void flatMap(String value, Collector<Tuple2<String, Integer>> out) {
            // Normalize and split the line into words
            String[] tokens = value.toLowerCase().split("\\W+");

            // Emit the pairs
            for (String token : tokens) {
                if (token.length() > 0) {
                    out.collect(new Tuple2<>(token, 1));
                }
            }
        }
    }
}
